File: scripts/pause.lua
-- Pause
--
-- Single source of truth for "is the game paused?". One call does it:
--
--   engine.setPaused(b) — flips the engine-side flag that gates
--   Unit/Thread.tickAllMovement (and anything else added later) AND,
--   in the same step, retimes the paused page's clock (World.Pause).
--   The page's chosen speed is captured when a pause epoch opens and
--   handed back when it closes, so a fast-forward (e.g. setTimeScale
--   10) survives a pause cycle and resumes at the speed the player
--   chose.
--
-- That pairing used to live HERE, and it could only ever hold for a
-- pause this module imposed (issue #1599). Several engine paths write
-- the pause flag without running a line of Lua — a notification
-- category with `pause: true` (Engine.PlayerEvent.Emit), the
-- acceptance of engine.saveWorld, a load publish — so the resume
-- branch was handing back whatever speed the last Lua-imposed pause
-- had recorded, 1.0 in an ordinary session. The engine now owns the
-- pair; this module owns the POLICY (who may pause, what a load
-- resumes at) and the player-facing logging.
--
-- Lua-side game-logic tick scripts (unit_ai, unit_resources,
-- building_spawn) check isPaused() at the top of their update() and
-- early-return. UI ticks (hud, debug, build_tool, tile_editor) stay
-- running so the player can pan, zoom, interact, and unpause.
--
-- Singleton via package.loaded so the engine.loadScript update tick
-- and the init.lua key handler see the same state.

local pause = package.loaded["scripts.pause"] or {}
package.loaded["scripts.pause"] = pause

pause.paused      = pause.paused      or false
-- Diagnostic MIRROR of the speed this module last saw a page paused
-- at, never the authority: the engine holds the real resume scale in
-- the paused page's own pause epoch, which is the only place that can
-- see an engine-imposed pause. Kept because it is the observable
-- tools/transactional_load_probe.py pins for "a rejected unpause
-- applied no time-scale side effect".
pause.prevTimeScale = pause.prevTimeScale or 1.0

function pause.isPaused()
    -- Defer to the engine flag rather than our local mirror. Avoids
    -- Lua/engine desync after a save load: the load handler flips
    -- enginePausedRef directly (auto-pause-on-save) without going
    -- through pause.set(), so our local `pause.paused` mirror would
    -- be stale. The engine flag is the source of truth; this
    -- module's pause.paused is now just a hint used during set/toggle
    -- to detect transitions for logging.
    return engine.isPaused()
end

function pause.set(b)
    b = b and true or false
    -- Check against the engine flag (source of truth), not the local
    -- mirror. The mirror can be stale after a path that flips
    -- enginePausedRef directly (a `pause: true` notification, the
    -- acceptance of a save, a load publish).
    if b == engine.isPaused() then
        -- Already in the requested state, so there is nothing to do —
        -- and in particular nothing to REPAIR. Re-pausing an already
        -- paused session used to have to heal a half-paused world here
        -- (ticks frozen, time-of-day still advancing) because those
        -- engine-side writers left the clock running; World.Pause now
        -- zeroes it as part of imposing the pause, so the split cannot
        -- arise. Re-pausing must never re-capture the clock either: it
        -- reads 0 during a pause, and storing that would resume the
        -- world stopped (#1599 requirement 4).
        return
    end
    -- engine.setPaused(false) can be REJECTED outright (issue #763
    -- round 15) while a load transaction is in flight -- staging runs
    -- before the save barrier's capture lock, so this call can land
    -- mid-transaction and resuming here could let the OLD, still-live
    -- session's simulation advance before the load either publishes or
    -- fails. Round 16 rereview: the engine reports whether it actually
    -- applied the flag, and this side must honour that. A rejection
    -- must leave EVERYTHING here untouched, matching the #763 "nothing
    -- changed" contract for the pre-load session -- pause.paused stays
    -- whatever it already was (matching the engine flag, which the
    -- rejection also left untouched), and so does prevTimeScale.
    --
    -- Read the live speed BEFORE asking the engine to pause, purely to
    -- keep the mirror above honest: by the time setPaused returns, the
    -- clock has already been zeroed. world.getTimeScale needs a pageId;
    -- with no world active (main menu, mid-transition) there is no
    -- clock to mirror and none for the engine to retime either. A
    -- RESUME reads nothing: the engine reinstates the speed it captured
    -- on the page that actually holds the pause epoch, which is not
    -- necessarily the page active by now.
    local liveScale = nil
    if b then
        local wid = world.getActiveWorldId()
        liveScale = wid and world.getTimeScale(wid) or nil
    end
    local applied = engine.setPaused(b)
    if applied == false then
        return
    end
    pause.paused = b
    if liveScale then
        pause.prevTimeScale = liveScale
    end
    engine.logInfo("Game " .. (b and "paused" or "resumed"))
end

function pause.toggle()
    -- Toggle off the engine flag, not the local mirror. See pause.set
    -- for why — the mirror can be stale after auto-pause-on-save.
    pause.set(not engine.isPaused())
end

-- Broadcast after the engine finishes a save-load, and the authoritative
-- statement of this module's LOAD policy: a load never restores a
-- pre-save speed. The published session comes up paused with every page
-- at the default 1.0 (time scale is deliberately not persisted), so the
-- speed asked for here is the default and nothing else -- never the
-- speed the world happened to be running at when it was saved, and
-- never whatever a compatibility blob from an older save carried.
--
-- world.setTimeScale on a PAUSED session records the request as the
-- pause epoch's resume scale and leaves the live clock at 0
-- (World.Thread.Command.Time), so this both keeps the loaded world
-- coherently paused and pins what resuming it runs at.
function pause.onSaveLoaded(survUnitIds, survBuildingIds)
    local wid = world.getActiveWorldId()
    if not wid then return end
    pause.prevTimeScale = 1.0
    world.setTimeScale(wid, 1.0)
end

-- Engine script hooks
function pause.init(scriptId)
    engine.logDebug("Pause module initializing...")
    -- No save registration (issue #761, requirement 5): pause is NOT
    -- persistent. `pause.paused` is never read for real logic (see
    -- `pause.isPaused`/`pause.set` above -- it's just a transition-
    -- detection hint compared against the authoritative
    -- `engine.isPaused()`), and a loaded session always begins paused
    -- and resumes at default speed via `pause.onSaveLoaded` above.
end

function pause.shutdown()
    engine.logDebug("Pause module shut down")
end

return pause

File: src/World/Pause.hs
{-# LANGUAGE Strict #-}
-- | The ONE writer of the (pause flag, paused page's clock) pair
--   (#1599).
--
--   @scripts\/pause.lua@ has always promised that a chosen fast-forward
--   \"survives a pause cycle and resumes at the speed the player
--   chose\". It could only keep that promise for a pause it imposed
--   itself: every engine-side writer of 'wsEnginePausedRef' — a
--   @pause: true@ notification category
--   ('Engine.PlayerEvent.Emit'), @engine.saveWorld@'s acceptance
--   ('Engine.Scripting.Lua.API.Save.acceptSaveRequest'), the world
--   thread's save re-assertion, a load publish — runs no Lua at all, so
--   the snapshot the Lua resume branch wrote back was whatever the last
--   Lua-imposed pause had left there (1.0 in an ordinary session).
--
--   Moving the pairing here removes the split rather than papering over
--   it: whoever imposes the pause captures the speed, so whoever lifts
--   it can give it back.
--
--   __An epoch records no OWNER, so the sources are counted
--   instead.__ Because a second pause landing on an already-paused
--   session is a complete no-op, nothing downstream could tell that a
--   notification still wanted the game paused when an autosave's
--   restore came to close the epoch — so the restore erased it, leaving
--   the popup that announced the pause standing over a running game
--   (#1730). 'imposePause' therefore bumps
--   'Engine.Core.State.enginePauseGenRef' on every call, no-op
--   included, while the two pauses a save imposes on its OWN behalf
--   ('reassertSavePause' and @acceptSaveRequest@'s 'imposePauseHeld')
--   deliberately do not — a save may not count itself as a reason to
--   decline its own restore. The player's pause is not counted here
--   either; @playerIntentGenRef@ already records it, and keeping the
--   two apart is what lets a declined restore name the real reason.
--
--   __A pause is an EPOCH, not a flag.__ The capture happens on the
--   unpaused→paused TRANSITION and only there. A second pause landing
--   while the session is already paused — the world thread re-asserting
--   a save's pause, a second notification, a save taken from an already
--   paused game — finds the clock already zeroed and must not replace
--   the captured speed with that zero, so 'imposePause' does nothing at
--   all in that case. That is what makes the epoch's speed survive an
--   arbitrary number of overlapping pause sources.
--
--   __The epoch belongs to a PAGE.__ It is stored in that page's own
--   'wsResumeScaleRef', so a resume restores the page whose clock was
--   actually zeroed rather than whichever page happens to be active
--   later, and a page that is gone by then takes its epoch with it
--   instead of donating its speed to a stranger.
--
--   __Opening and closing an epoch are LINEARIZABLE against each
--   other.__ An epoch is three writes (the flag, the page's captured
--   speed, the page's clock), so protecting only the flag is not enough:
--   a notification opening one on the world thread could publish the
--   flag, have the Lua thread observe it, close the epoch it cannot yet
--   see and clear the flag, and then finish capturing — leaving an
--   UNPAUSED session at scale 0 with a resume slot nobody will ever
--   consume. Every transition therefore runs under
--   'Engine.Core.Capability.WorldSim.withPlayerIntentHeld', the mutex
--   the player's own pause and time-scale transitions already take
--   (holding it, never bumping it: an engine-imposed pause is not player
--   intent, #913). Callers that ALREADY hold that lock —
--   @engine.setPaused@, @engine.saveWorld@'s acceptance, an autosave's
--   conditional restore — must use the @…Held@ variants instead, since
--   the lock is a plain 'Control.Concurrent.MVar.MVar' and re-entering
--   it deadlocks.
module World.Pause
    ( imposePause
    , imposePauseHeld
    , reassertSavePause
    , beginPauseEpoch
    , releasePause
    , releasePauseHeld
    , setPauseResumeScale
    ) where

import UPrelude
import Data.IORef (readIORef, writeIORef, atomicModifyIORef')
import Engine.Core.Capability.WorldSim
    ( WorldSimCapability(wsEnginePausedRef, wsEnginePauseGenRef
                        , wsWorldManagerRef)
    , withPlayerIntentHeld )
import World.Types
    ( WorldManager(wmWorlds), WorldState(wsTimeScaleRef, wsResumeScaleRef)
    , visiblePageState )

-- | Pause the session, starting a pause epoch if one is not already
--   running.
--
--   On the unpaused→paused transition the visible page's chosen speed
--   is captured into its 'wsResumeScaleRef' and its clock zeroed, as
--   one step. When the session was ALREADY paused this is a complete
--   no-op — see the module haddock: the running epoch's captured speed
--   is the player's real one and the clock it would re-capture is
--   already zero.
--
--   Every engine-side pause writer goes through here, through
--   'imposePauseHeld', or — for a save re-asserting its own pause —
--   through 'reassertSavePause'.
--
--   This one also COUNTS the assertion (#1730): the caller is a source
--   independent of any save in flight, so an autosave that finds the
--   count moved since its acceptance leaves the pause alone rather than
--   erasing someone else's.
imposePause ∷ WorldSimCapability → IO ()
imposePause wsc = withEpochLock wsc $ do
    countEnginePauseAssertion wsc
    imposePauseHeld wsc

-- | The pause a SAVE re-asserts on its own behalf, once its queued
--   @WorldSave@ reaches the world thread
--   ('World.Thread.Command.Save.WriteWorld.handleWorldSaveCommand').
--
--   Identical to 'imposePause' in what it does to the epoch — almost
--   always nothing, since the Lua acceptance already opened it — and
--   deliberately different in what it RECORDS: it does not count as an
--   independent pause assertion (#1730). A save asserting its own pause
--   a second time must never be mistaken for someone else wanting the
--   game paused, or every autosave would decline to restore the state
--   it just took.
reassertSavePause ∷ WorldSimCapability → IO ()
reassertSavePause wsc = withEpochLock wsc (imposePauseHeld wsc)

-- | Record that an engine pause source INDEPENDENT of any running save
--   has asserted a pause (#1730).
--
--   Counted on every call, including the overwhelmingly common one that
--   finds the session already paused and changes nothing: that no-op is
--   precisely the case the counter exists to make visible, because the
--   epoch itself records no owner. Over-counting is the safe direction —
--   it can only make an autosave decline to resume a game somebody else
--   wanted paused — and under-counting is not.
--
--   Bare 'atomicModifyIORef'' rather than a lock of its own: every
--   caller reaches this through 'withEpochLock', and the two sites that
--   snapshot and compare the counter take that same mutex.
countEnginePauseAssertion ∷ WorldSimCapability → IO ()
countEnginePauseAssertion wsc =
    atomicModifyIORef' (wsEnginePauseGenRef wsc) (\g → (g + 1, ()))

-- | 'imposePause' for a caller already holding the player-intent lock.
--
--   The epoch half ALONE: it does not count an engine pause assertion
--   (#1730), because its two callers are the two that must not — a
--   save's own acceptance
--   ('Engine.Scripting.Lua.API.Save.acceptSaveRequest') and the
--   player's @engine.setPaused@, whose intent @playerIntentGenRef@
--   already records.
imposePauseHeld ∷ WorldSimCapability → IO ()
imposePauseHeld wsc = do
    wasPaused ← atomicModifyIORef' (wsEnginePausedRef wsc) (\p → (True, p))
    unless wasPaused $ captureVisibleClock wsc

-- | Start a FRESH pause epoch, whatever the flag already said.
--
--   For 'World.Load.Publish' alone: a load transaction pauses the
--   OUTGOING session at acceptance, so by the time the replacement
--   session is live the flag is already set and 'imposePause' would
--   preserve an epoch belonging to pages that no longer exist. The
--   published session gets its own epoch instead — its active page's
--   clock zeroed and the default speed every loaded page comes up at
--   captured — which is the load policy @scripts\/pause.lua@'s
--   @onSaveLoaded@ states: a load resumes at 1.0, never at some
--   pre-save speed.
--
--   Uncounted (#1730), like 'imposePauseHeld': a load and a save are
--   mutually exclusive for the load's whole duration
--   ('Engine.Scripting.Lua.API.Save.loadSaveFn' rejects one outright
--   while the other runs), so there is no autosave window for this to
--   land in — and the session it opens the epoch over is not the one
--   any earlier request was taken from.
beginPauseEpoch ∷ WorldSimCapability → IO ()
beginPauseEpoch wsc = withEpochLock wsc $ do
    writeIORef (wsEnginePausedRef wsc) True
    captureVisibleClock wsc

-- | Hold the mutex every epoch transition shares — see the module
--   haddock for why the three writes have to be one critical section.
--   The generation is deliberately NOT advanced: these are the engine's
--   own writes, and #913 reserves a bump for the player's.
withEpochLock ∷ WorldSimCapability → IO α → IO α
withEpochLock wsc act = withPlayerIntentHeld wsc (const act)

-- | Capture the visible page's live scale as this epoch's resume scale
--   and zero its clock, in that order — the clock must never be read as
--   zeroed before the value it was zeroed from has been recorded.
captureVisibleClock ∷ WorldSimCapability → IO ()
captureVisibleClock wsc = do
    mgr ← readIORef (wsWorldManagerRef wsc)
    forM_ (visiblePageState mgr) $ \ws → do
        scale ← readIORef (wsTimeScaleRef ws)
        writeIORef (wsResumeScaleRef ws) (Just scale)
        writeIORef (wsTimeScaleRef ws) 0

-- | End the pause epoch: give every page carrying a resume scale its
--   speed back, then clear the flag.
--
--   Scales first, flag last, so the pair is never momentarily readable
--   as \"running at scale 0\". Only pages whose clock this epoch
--   actually took hold of carry a value, so this cannot write a scale
--   onto a bystander page; running it on an already-unpaused session
--   finds nothing to restore and is a no-op.
releasePause ∷ WorldSimCapability → IO ()
releasePause wsc = withEpochLock wsc (releasePauseHeld wsc)

-- | 'releasePause' for a caller already holding the player-intent lock.
releasePauseHeld ∷ WorldSimCapability → IO ()
releasePauseHeld wsc = do
    mgr ← readIORef (wsWorldManagerRef wsc)
    forM_ (wmWorlds mgr) $ \(_, ws) → do
        mScale ← takeResumeScale (wsResumeScaleRef ws)
        forM_ mScale $ writeIORef (wsTimeScaleRef ws)
    writeIORef (wsEnginePausedRef wsc) False
  where
    takeResumeScale ref = atomicModifyIORef' ref (\m → (Nothing, m))

-- | Record the speed a page should come back at when the current epoch
--   ends, without touching its live clock.
--
--   Two callers, both of which know a resume scale the capture could
--   not have seen: @world.setTimeScale@ landing while the session is
--   paused (the player's choice would otherwise be dropped, since a
--   running scale must never be stored on a paused page), and a
--   successful autosave handing back the pre-request speed its own
--   'World.Save.Types.AutosaveRequest' recorded.
--
--   Takes no lock of its own: both callers already hold the epoch mutex,
--   because in both cases this write and the flag read (or the release)
--   beside it have to be one step.
setPauseResumeScale ∷ WorldState → Float → IO ()
setPauseResumeScale ws scale = writeIORef (wsResumeScaleRef ws) (Just scale)
